cluster-autoscaler/cloudprovider/aws/api/instance_info.go

```diff
@@ -0,0 +1,356 @@
+	return &instanceInfoService{
+		client: client,
+		cache:  make(instanceInfoCache),
+	}
+}
+
+type instanceInfoService struct {
+	client httpClient
+	cache  instanceInfoCache
+	sync.RWMutex
+}
+
+// DescribeInstanceInfo returns the corresponding aws instance info by given instance type and availability zone.
+func (s *instanceInfoService) DescribeInstanceInfo(instanceType string, availabilityZone string) (*InstanceInfo, error) {
+	if s.shouldSync(availabilityZone) {
+		if err := s.sync(availabilityZone); err != nil {
+			// TODO <mrcrgl> may this be tolerated for resilience
+			return nil, fmt.Errorf("failed to sync aws product and price information: %v", err)
+		}
+	}
+
+	if bucket, found := s.cache[availabilityZone]; found {
+		for _, info := range bucket.info {
+			if info.InstanceType == instanceType {
+				return &info, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("instance info not available for instance type %s in zone %s", instanceType, availabilityZone)
+}
+
+func (s *instanceInfoService) shouldSync(availabilityZone string) bool {
+	bucket, found := s.cache[availabilityZone]
+	if !found {
+		return true
+	}
+
+	return bucket.LastSync().Before(time.Now().Truncate(instanceInfoCacheMaxAge))
+}
+
+func (s *instanceInfoService) sync(availabilityZone string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	start := time.Now()
+
+	bucket, found := s.cache[availabilityZone]
+	if !found {
+		bucket = new(regionalInstanceInfoBucket)
+		s.cache[availabilityZone] = bucket
+	}
+
+	response, err := s.fetch(availabilityZone, bucket.ETag)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		glog.V(4).Infof("Synchronized aws ec2 instance information for availability zone %s - took %s", availabilityZone, time.Now().Sub(start).String())
+	}()
+
+	if response == nil {
+		bucket.SetLastSync()
+		return nil
+	}
+
+	instances := make([]InstanceInfo, 0)
+	now := time.Now()
+
+	for _, product := range response.Products {
+		sku := product.SKU
+		attr := product.Attributes
+
+		// TODO <mrcrgl> find better solution for the case of windows installations for instance.
+		if attr.OperatingSystem != instanceOperatingSystemLinux {
+			continue
+		}
+
+		// We do actually only support Shared tenancy instances.
+		// See for more information: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-purchasing-options.html
+		if attr.Tenancy != instanceTenancyShared {
+			continue
+		}
+
+		if len(attr.InstanceType) == 0 {
+			continue
+		}
+
+		i := InstanceInfo{
+			InstanceType: attr.InstanceType,
+		}
+
+		var err error
+		if attr.Memory != "" && attr.Memory != "NA" {
+			if i.MemoryMb, err = parseMemory(attr.Memory); err != nil {
+				return fmt.Errorf("parser error %v", err)
+			}
+		}
+
+		if attr.VCPU != "" {
+			if i.VCPU, err = parseCPU(attr.VCPU); err != nil {
+				return fmt.Errorf("parser error %v", err)
+			}
+		}
+		if attr.GPU != "" {
+			if i.GPU, err = parseCPU(attr.GPU); err != nil {
+				return fmt.Errorf("parser error %v", err)
+			}
+		}
+
```

Comment:
Since this json file is so large, perhaps you could probably make these searching/filtering a bit more efficient. Perhaps using maps. One pass to gather all the sku and another pass to get the pricing data.

Reply:
But better to use the AWS API to get prices like it was suggested earlier. Much better than parsing a big json file.